util/mutexlock.h

```diff
@@ -132,4 +132,30 @@ class SpinMutex {
   std::atomic<bool> locked_;
 };
 
+//
+// Inspired by Guava: https://github.com/google/guava/wiki/StripedExplained
+// A striped Lock. This offers the underlying lock striping similar
+// to that of ConcurrentHashMap in a reusable form, and extends it for
+// semaphores and read-write locks. Conceptually, lock striping is the technique
+// of dividing a lock into many <i>stripes</i>, increasing the granularity of a
+// single lock and allowing independent operations to lock different stripes and
+// proceed concurrently, instead of creating contention for a single lock.
+//
+template <class T, class P>
+class Striped {
+ public:
+  Striped(size_t stripes, std::function<uint64_t(const P &)> hash)
+      : locks_(stripes), hash_(hash) {}
+
+  T *get(const P &key) {
+    uint64_t h = hash_(key);
+    size_t index = h % locks_.size();
+    return &locks_[index];
+  }
+
+ private:
+  std::vector<T> locks_;
```

Comment:
Can we put those locks in different cache lines to prevent false sharing? We do something like this in other places: https://github.com/facebook/rocksdb/blob/fdf882ded218344c136c97daf76dfb59e4bc155f/monitoring/statistics.h#L83

Reply:
Please review

Reply:
OK I forgot destructor. Sorry...

Reply:
Fixed